File: mobile/ios/OpenU60/Features/RouterSettings/Telemetry/TelemetryBlockerViewModel.swift
import SwiftUI

@Observable
@MainActor
final class TelemetryBlockerViewModel {
    var filterConfig: DomainFilterConfig = .empty
    var isLoading: Bool = false
    var message: String?
    var messageIsError: Bool = false
    var newDomain: String = ""

    private let client: AgentClient
    private let authManager: AuthManager

    init(client: AgentClient, authManager: AuthManager) {
        self.client = client
        self.authManager = authManager
    }

    func refresh() async {
        isLoading = true
        message = nil

        do {
            let data = try await client.getJSON("/api/router/domain-filter")
            filterConfig = TelemetryParser.parseDomainFilter(data)
        } catch {
            showMessage("加载过滤规则失败：\(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func toggleFilter(enabled: Bool) async {
        isLoading = true

        do {
            let _ = try await client.putJSON("/api/router/domain-filter", body: ["enable": enabled ? "1" : "0"])
            showMessage(enabled ? "域名过滤已启用" : "域名过滤已禁用", isError: false)
            filterConfig.enabled = enabled
        } catch {
            showMessage("失败：\(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func addDomain(_ domain: String) async {
        let trimmed = domain.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            showMessage("请输入域名", isError: true)
            return
        }

        isLoading = true

        do {
            let _ = try await client.putJSON("/api/router/domain-filter", body: [
                "action": "add",
                "domain": trimmed,
                "enabled": "1"
            ])
            newDomain = ""
            showMessage("已添加 \(trimmed)", isError: false)
            filterConfig.rules.append(DomainFilterRule(id: UUID().uuidString, domain: trimmed, enabled: true))
        } catch {
            showMessage("失败：\(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func removeDomain(_ rule: DomainFilterRule) async {
        isLoading = true

        do {
            let _ = try await client.putJSON("/api/router/domain-filter", body: [
                "action": "delete",
                "id": rule.id
            ])
            showMessage("已移除 \(rule.domain)", isError: false)
            filterConfig.rules.removeAll { $0.id == rule.id }
        } catch {
            showMessage("失败：\(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func blockAllTelemetry() async {
        isLoading = true
        let existingDomains = Set(filterConfig.rules.map(\.domain))

        var added = 0
        for domain in TelemetryParser.knownTelemetryDomains {
            guard !existingDomains.contains(domain) else { continue }
            do {
                let _ = try await client.putJSON("/api/router/domain-filter", body: [
                    "action": "add",
                    "domain": domain,
                    "enabled": "1"
                ])
                added += 1
                filterConfig.rules.append(DomainFilterRule(id: UUID().uuidString, domain: domain, enabled: true))
            } catch {
                // Continue with remaining domains
            }
        }

        if added > 0 {
            showMessage("已屏蔽 \(added) 个遥测域名", isError: false)
        } else {
            showMessage("所有遥测域名均已被屏蔽", isError: false)
        }

        isLoading = false
    }

    private func showMessage(_ text: String, isError: Bool) {
        message = text
        messageIsError = isError
    }
}

File: mobile/android/OpenU60/app/src/main/java/com/openu60/feature/router/telemetry/TelemetryBlockerViewModel.kt
package com.openu60.feature.router.telemetry

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.openu60.core.model.DomainFilterConfig
import com.openu60.core.model.TelemetryParser
import com.openu60.core.network.AgentClient
import com.openu60.core.network.AgentError
import com.openu60.core.network.AuthManager
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class TelemetryBlockerState(
    val config: DomainFilterConfig = DomainFilterConfig.empty,
    val isLoading: Boolean = false,
    val message: String? = null,
    val messageIsError: Boolean = false,
    val newDomain: String = "",
)

@HiltViewModel
class TelemetryBlockerViewModel @Inject constructor(
    private val agentClient: AgentClient,
    private val authManager: AuthManager,
) : ViewModel() {

    private val _state = MutableStateFlow(TelemetryBlockerState())
    val state: StateFlow<TelemetryBlockerState> = _state.asStateFlow()

    fun refresh() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val data = agentClient.getJSON("/api/router/domain-filter")
                val config = TelemetryParser.parseDomainFilter(data)
                _state.value = _state.value.copy(config = config, isLoading = false)
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) refresh() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun toggleFilter(enabled: Boolean) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.putJSON("/api/router/domain-filter", mapOf("enable" to if (enabled) "1" else "0"))
                _state.value = _state.value.copy(message = if (enabled) "域名过滤已启用" else "域名过滤已禁用", messageIsError = false)
                refresh()
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) toggleFilter(enabled) else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun updateNewDomain(value: String) {
        _state.value = _state.value.copy(newDomain = value)
    }

    fun addRule(domain: String) {
        val trimmed = domain.trim()
        if (trimmed.isEmpty()) {
            setError("请输入域名")
            return
        }
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.putJSON(
                    "/api/router/domain-filter",
                    mapOf(
                        "action" to "add",
                        "domain" to trimmed,
                        "enabled" to "1",
                    ),
                )
                val config = TelemetryParser.parseDomainFilter(agentClient.getJSON("/api/router/domain-filter"))
                _state.value = _state.value.copy(
                    config = config,
                    isLoading = false,
                    message = "已添加 $trimmed",
                    messageIsError = false,
                    newDomain = "",
                )
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) addRule(trimmed) else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun removeRule(id: String) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.putJSON(
                    "/api/router/domain-filter",
                    mapOf(
                        "action" to "delete",
                        "id" to id,
                    ),
                )
                val config = TelemetryParser.parseDomainFilter(agentClient.getJSON("/api/router/domain-filter"))
                _state.value = _state.value.copy(
                    config = config,
                    isLoading = false,
                    message = "规则已移除",
                    messageIsError = false,
                )
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) removeRule(id) else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun blockAllTelemetry() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val existing = _state.value.config.rules.map { it.domain.lowercase() }.toSet()
                var added = 0
                for (domain in TelemetryParser.knownTelemetryDomains) {
                    if (domain.lowercase() in existing) continue
                    agentClient.putJSON(
                        "/api/router/domain-filter",
                        mapOf(
                            "action" to "add",
                            "domain" to domain,
                            "enabled" to "1",
                        ),
                    )
                    added += 1
                }
                val config = TelemetryParser.parseDomainFilter(agentClient.getJSON("/api/router/domain-filter"))
                _state.value = _state.value.copy(
                    config = config,
                    isLoading = false,
                    message = if (added > 0) "已屏蔽 $added 个遥测域名" else "所有遥测域名均已存在",
                    messageIsError = false,
                )
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) blockAllTelemetry() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    private fun setError(msg: String?) {
        _state.value = _state.value.copy(isLoading = false, message = msg ?: "未知错误", messageIsError = true)
    }
}
